rewrite-core/src/main/java/org/openrewrite/ShowGitProvenance.java

```diff
@@ -0,0 +1,36 @@
+package org.openrewrite;
+
+import org.openrewrite.internal.lang.Nullable;
+import org.openrewrite.marker.GitProvenance;
+import org.openrewrite.marker.Markup;
+
+public class ShowGitProvenance extends Recipe {
+
+    @Override
+    public String getDisplayName() {
+        return "Show `GitProvenance` in files";
+    }
+
+    @Override
+    public String getDescription() {
+        return "List out the contents of the `GitProvenance` markers in each file under source control.";
+    }
+
+    @Override
+    protected TreeVisitor<?, ExecutionContext> getVisitor() {
+        return new TreeVisitor<Tree, ExecutionContext>() {
+            @Override
+            public @Nullable Tree visitSourceFile(SourceFile sourceFile, ExecutionContext executionContext) {
+                Tree visited = super.visitSourceFile(sourceFile, executionContext);
+                if (visited == null) {
+                    return null;
+                }
+                return visited.getMarkers()
+                        .findFirst(GitProvenance.class)
+                        .map(gitProvenance ->
+                                Markup.info(visited, "GitProvenance found", gitProvenance.toString()))
+                        .orElse(visited);
+            }
+        };
+    }
+}
```

Comment:
Every single file parsed from a git repository is given a `GitProvenance` marker, so as written your recipe will have results for every single file. I suggest changing it so that each _unique_ `GitProvenance` marker is applied only once. So you'll see if, somehow, there is more than one. And in all other cases it will be easy to see that there is _exactly_ one, as you expect. 

```suggestion
package org.openrewrite;

import org.openrewrite.internal.ListUtils;
import org.openrewrite.marker.GitProvenance;
import org.openrewrite.marker.Markup;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class ShowGitProvenance extends Recipe {

    @Override
    public String getDisplayName() {
        return "Show git source control metadata";
    }

    @Override
    public String getDescription() {
        return "List out the contents of each unique `GitProvenance` marker in the set of source files. " +
                "When everything is working correctly exactly one such marker should be printed as all source files are " +
                "expected to come from the same repository / branch / commit ID.";
    }

    @Override
    protected List<SourceFile> visit(List<SourceFile> before, ExecutionContext ctx) {
        Set<GitProvenance> seenProvenance = new HashSet<>();
        return ListUtils.map(before, sourceFile -> {
            GitProvenance provenance = sourceFile.getMarkers().findFirst(GitProvenance.class).orElse(null);
            if(provenance == null) {
                return sourceFile;
            }
            if(seenProvenance.contains(provenance)) {
                return sourceFile;
            }
            seenProvenance.add(provenance);
            return Markup.info(sourceFile, "GitProvenance: origin: " + provenance.getOrigin() + ", branch: "
                    + provenance.getBranch() + ", commit: " + provenance.getChange());
        });
    }
}

```